src/h5web/toolbar/Toolbar.tsx

```diff
@@ -1,32 +1,34 @@
-import { ReactElement, Children } from 'react';
-import { useMeasure } from '@react-hookz/web';
-import { useList } from 'react-use';
+import type { ReactElement, ReactNode } from 'react';
+import { useMap, useMeasure } from '@react-hookz/web';
 import styles from './Toolbar.module.css';
 import Separator from './Separator';
 import OverflowMenu from './OverflowMenu';
 import MeasuredControl from './MeasuredControl';
+import flattenChildren from 'react-keyed-flatten-children';
+import { isReactElement } from '../guards';
 
 interface Props {
-  children?: ReactElement | (ReactElement | undefined)[];
+  children: ReactNode;
 }
 
 function Toolbar(props: Props) {
   const { children } = props;
 
-  // Convert `children` to array, as it cannot be used directly (and remove `undefined` children)
-  const allChildren = Children.toArray(children) as ReactElement[];
+  /* Convert `children` to flat array by traversing nested arrays and fragments.
+   * (Note that `flattenChildren` guarantees stable string keys regardless of JSX logic.) */
+  const allChildren = flattenChildren(children).filter(isReactElement);
 
   const [containerSize, containerRef] = useMeasure<HTMLDivElement>();
   const availableWidth = containerSize ? containerSize.width : 0;
 
-  const [childrenWidths, { updateAt: setChildWidth }] = useList<number>();
+  const childrenWidths = useMap<string, number>();
 
   // Group visible and hidden children based on their accumulated width
   const [visibleChildren, hiddenChildren, allMeasured] = allChildren.reduce<
     [ReactElement[], ReactElement[], boolean, number]
   >(
-    ([accVisible, accHidden, accAllMeasured, accWidth], child, index) => {
-      const width = childrenWidths[index] ?? 0;
+    ([accVisible, accHidden, accAllMeasured, accWidth], child) => {
+      const width = childrenWidths.get(child.key as string) ?? 0;
```

Comment:
As mentioned, I've confirmed that all children elements have stable, string keys thanks to `flattenChildren` (though I think this was also the case with `React.Children.toArray`...), so I can store the widths against those keys. Note that `as string` is required here because `ReactElement.key` has type `string | number | null`...

Reply:
Would it be safer to `assertString` instead of `as string` ? Or do you trust `flattenChildren` enough ?

Reply:
I trust it enough. Also, if it's anything else, then the `get` will return `undefined` anyway.